src/main/java/jalf/relation/algebra/Project.java

```diff
@@ -59,4 +67,28 @@ public AttrList getAttributes() {
         return visitor.visit(this);
     }
 
+
+    /**
+     * 4 case to check
+     * - Ps intersect Kx = Kx : Kn = Kx
+     * - Ps intersect Kx = Ps : Kn = on
+     * - Ps intersect Kx = null : Kn = on
+     * - Ps intersect Kx = sx : Kn = on
+     */
+    @Override
+    public Keys lazyComputeKey(){
+        Keys keys=operand.getKeys();
+
+        Predicate<Key> P= (k)-> k.intersect(this.attributes).equals(k.toAttrList());
+        Set<Key> keyfilter=keys.stream().filter(P).collect(Collectors.toSet());
+        if( keyfilter.isEmpty()){
+            // no key is good
+            return  new Keys(Key.candidate(this.attributes));
+        }
+        else{
+            //on a filtrer les mauvaises clefs
+            return new Keys(keyfilter);
+        }
+    }
```

Comment:
Rephrase and document in terms of new `isKeyPreserving` / `isKeyPreserved` abstractions (methods on Project):

```
isKeyPreserving(Keys keys) -> Boolean
isKeyPreserved(Key key) -> Boolean
```